cpp/src/arrow/util/decimal.cc

```diff
@@ -799,7 +800,7 @@ Status Decimal128::ToArrowStatus(DecimalStatus dstatus) const {
   return arrow::ToArrowStatus(dstatus, 128);
 }
 
-std::ostream& operator<<(std::ostream& os, const Decimal128& decimal) {
+ARROW_EXPORT std::ostream& operator<<(std::ostream& os, const Decimal128& decimal) {
```

Comment:
Can we avoid using `ARROW_EXPORT` in `.cc`?

Does this work instead of adding `ARROW_EXPORT` in `.cc`?

```diff
diff --git a/cpp/src/arrow/util/visibility.h b/cpp/src/arrow/util/visibility.h
index 1498d2085a..bad80dd8a2 100644
--- a/cpp/src/arrow/util/visibility.h
+++ b/cpp/src/arrow/util/visibility.h
@@ -77,8 +77,8 @@
 #endif
 #endif
 
-#define ARROW_FRIEND_EXPORT
-#define ARROW_TEMPLATE_EXPORT
+#define ARROW_FRIEND_EXPORT ARROW_EXPORT
+#define ARROW_TEMPLATE_EXPORT ARROW_EXPORT
 
 // [[gnu::visibility("default")]] even when #included by a non-arrow source
 #define ARROW_FORCE_EXPORT [[gnu::visibility("default")]]
```

Reply:
Makes sense - yep that appears to work as well

Reply:
Doesn't look like clang likes that. From what I understand, putting the attribute next to a friend declaration won't work - it needs to be attached to the definition:

https://stackoverflow.com/a/52332379/621736

Should I move the definition into the header file or should we revert back to the original design?

Reply:
Or we may be able to change the macro back to using gcc attribute syntax instead of the C++ attribute (haven't tested - just a thought)

Reply:
Oh, sorry. Let's use your original approach not my suggestion.